crucible/src/Lang/Crucible/CFG/Transforms.hs

```diff
@@ -0,0 +1,730 @@
+-----------------------------------------------------------------------
+-- |
+-- Module           : Lang.Crucible.CFG.Transform
+-- Description      : Provides transformations on pre-SSA CFGs
+-- Copyright        : (c) Galois, Inc 2020
+-- License          : BSD3
+-- Maintainer       : 
+-- Stability        : experimental
+--
+-- This modules defines transformations on CFGs that feature mutable
+-- registers.
+------------------------------------------------------------------------
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE ViewPatterns #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TupleSections #-}
+module Lang.Crucible.CFG.Transforms
+  ( earlyMergeLoops
+  ) where
+
+import           Control.Monad ((>=>))
+import           Control.Applicative ((<**>))
+import qualified Data.Graph.Inductive as G
+import qualified Data.Foldable as Fold
+import           Data.List (nub, sort)
+import qualified Data.Map.Strict as Map
+import           Data.Maybe (fromMaybe)
+import           Data.Parameterized.Classes
+import qualified Data.Parameterized.Context as Ctx
+import qualified Data.Parameterized.Map as MapF
+import           Data.Parameterized.Nonce
+import           Data.Parameterized.Some
+import           Data.Parameterized.TraversableFC
+import           Data.Set (Set)
+import qualified Data.Set as Set
+import           Data.Sequence (Seq)
+import qualified Data.Sequence as Seq ((<|), fromList)
+import           Data.String (fromString)
+
+import           What4.ProgramLoc (Position(..), Posd(..))
+
+import           Lang.Crucible.CFG.Expr
+import           Lang.Crucible.CFG.Reg
+import           Lang.Crucible.Types
+import Data.Kind
+
+-- | Note: A footer is a block that has *any* successor whose must-set of successors are a dominator
+-- of the footer. The header is this dominator.
+-- i.e. if:
+-- h -> b -> f -> e
+-- ^          `-> c
+--  \____________/
+-- then f is the footer
+data LoopInfo s = LoopInfo
+  { liFooter     :: !(BlockID s) -- ^ This is the first block with a backedge "chain"
+  , liHeader     :: !(BlockID s)
+  , liMembers    :: !(Set (BlockID s))
+  , liEarlyExits :: ![CFGEdge s]
+  , liFooterIn   :: ![CFGEdge s]
+  , liDominators :: ![BlockID s]
+  } deriving (Eq, Show)
+type CFGEdge s = (BlockID s, BlockID s)
+
+instance Ord (LoopInfo s) where
+   li1 <= li2 = isNested li1 li2
+
+-- | Detect all loops in a cfg.
+-- The assumption is that two backedges in a cfg will have distinct destination blocks.
```

Comment:
You're right, I'm not sure that it should be in general. While I'm not convinced it will matter in practice if we discard that case (i.e., don't apply the transformation), it might be worth double checking if this assumption is critical. For now, I think this assumption ought to be checked.